storage/range.go

```diff
@@ -819,23 +852,37 @@ func (r *Range) applyRaftCommand(ctx context.Context, index uint64, originNodeID
 			if log.V(1) {
 				log.Infoc(ctx, "found response cache entry for %+v", args.Header().CmdID)
 			}
-			return reply.Header().GoError()
+			return nil, err
 		} else if ok && err != nil {
-			newErr := util.Errorf("unable to read result for %+v from the response cache: %s", args, err)
-			reply.Header().SetGoError(newErr)
-			return newErr
+			return nil, newReplicaCorruptionError(
+				util.Errorf("could not read from response cache"), err)
 		}
 	}
 
-	// Verify the leader lease is held; Note that we don't require the
-	// leader lease when trying to grant the leader lease!
-	if _, ok := args.(*proto.InternalLeaderLeaseRequest); !ok {
-		if !r.verifyLeaderLease(originNodeID, header.Timestamp) {
-			err := r.newNotLeaderError()
-			reply.Header().SetGoError(err)
-			return err
+	committed := false
+
+	defer func() {
+		if !committed {
+			// We didn't commit the batch, but advance the last applied index nonetheless.
+			if sErr := setAppliedIndex(r.rm.Engine(), r.Desc().RaftID, index); sErr != nil {
+				rErr = newReplicaCorruptionError(
+					util.Errorf("could not advance applied index"), sErr, rErr)
+				return
+			}
+			atomic.StoreUint64(&r.appliedIndex, index)
 		}
-	}
+		if proto.IsWrite(args) {
+			// No matter the result, add result to the response cache if this
```

Comment:
The error is a part of the response and must be saved in the response cache. We must do `Reply.Header().SetGoError(rError)` here.